### ISSUE TITLE
Bot becomes stuck with error "Error handling sync Error: Expect value to be String, but received Undefined"

### ISSUE THREAD

**Issue by stronk7** [turt2live/matrix-bot-sdk#256]
**Describe the bug**
After upgrading to v0.6.1 the bot becomes unresponsive after few time (days). The very same code, using previous v0.6.0-beta.8 seems to be unaffected.

Note the time (days) to get the problem is not always the same, we have reproduced it here after 2 days, after 5 days, after 7 days... the exact moment when it starts happening can be detected in logs because they start logging the `Error handling sync Error: Expect value to be String, but received Undefined` message every few seconds (each sync attempt, maybe?).

Once that happens the bot doesn't work anymore, though the node process continues running apparently.

**To Reproduce**
1. Start a v0.6.1 based bot.
2. Ensure it's a member of a varietè of rooms (normal and e2e).
3. Wait an undefined amount of time (I've not been able to find any cause/action causing the problem to heppen).
4. The bot stops working (with the logs details added below).

**Expected behavior**
The bot should continue working, behaving better when that situation is found.

**Log snippet**
```
...
Mon, 22 Aug 2022 14:25:26 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Mon, 22 Aug 2022 14:25:26 GMT [INFO] [MatrixClientLite] Backing off for 5162.254702188716ms

Mon, 22 Aug 2022 14:25:31 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Mon, 22 Aug 2022 14:25:31 GMT [INFO] [MatrixClientLite] Backing off for 6853.857301967255ms

Mon, 22 Aug 2022 14:25:39 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Mon, 22 Aug 2022 14:25:39 GMT [INFO] [MatrixClientLite] Backing off for 12180.066837999479ms

Mon, 22 Aug 2022 14:25:51 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Mon, 22 Aug 2022 14:25:51 GMT [INFO] [MatrixClientLite] Backing off for 6209.698348785948ms
...
```

**Additional context**
Server we are using is synapse:v1.57.1, right now.

**Comment by stronk7** [comment]
A little update on this, I've had v0.6.2 here in out staging environment up and running ok perfectly. All unencrypted rooms.

But then, I opened my 1:1 chat with the bot (that is encrypted) and sent to it a simple `!ping` and then the log entries shared above started immediately and the bot has stopped working everywhere.

Also, not sure if that's relevant, in my 1:1: chat I got from Element a (sorry this is not exact, it went away before copying it) message saying "message could not be decrypted, click to ask for the keys of all your sessions" (aprox!).

But the point is that, as soon as I accessed to an encrypted (the 1:1) room, has stopped working and the logs are, all the time, showing the messages above.

I'm going to restart it again to see if can reproduce it consistently...

Ciao :-)

**Comment by stronk7** [comment]
Yeah, confirmed. New instance created, I send a couple of `!ping` in the 1:1 encrypted room and in a few seconds, the logging started and it's not responsive anymore. Here there are the complete logs since startup (note that some are own implementation).

```
Already up to date.
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Sleeping for 60 seconds]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Using './config-dev.yml' for configuration (context: 'dev')]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Using './storage-dev' for storage (context: 'dev')]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Using already existing token]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Starting HTTP server]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Starting database pool]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Starting profiles cache]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Setting Jenkins webhook]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Checking chat logs database structures]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Server started at http://localhost:80]
Fri, 30 Sep 2022 14:54:49 GMT [INFO] [Setting chat logs webhook]
Fri, 30 Sep 2022 14:54:49 GMT [ERROR] [MatrixHttpClient] (REQ-4) { errcode: 'M_NOT_FOUND', error: 'Event not found.' }
Fri, 30 Sep 2022 14:54:49 GMT [ERROR] [MatrixHttpClient] (REQ-5) { errcode: 'M_NOT_FOUND', error: 'Event not found.' }
Fri, 30 Sep 2022 14:54:50 GMT [ERROR] [MatrixHttpClient] (REQ-6) { errcode: 'M_NOT_FOUND', error: 'Event not found.' }
Fri, 30 Sep 2022 14:54:50 GMT [INFO] [MatrixClientLite] End-to-end encryption enabled
Fri, 30 Sep 2022 14:54:50 GMT [INFO] [Bot started!]
Fri, 30 Sep 2022 14:55:17 GMT [ERROR] [MatrixHttpClient] (REQ-13) { errcode: 'M_NOT_FOUND', error: 'Event not found.' }
Fri, 30 Sep 2022 14:55:17 GMT [INFO] [Jira linking: Ignored user @ !VlyOdAIoWKBBMwwrMU:moodle.com: @nextmoodlebot:moodle.com said 'Hey @stronk7:moodle.com!']
Fri, 30 Sep 2022 14:55:31 GMT [ERROR] [MatrixHttpClient] (REQ-21) { errcode: 'M_NOT_FOUND', error: 'Event not found.' }
Fri, 30 Sep 2022 14:55:32 GMT [INFO] [Jira linking: Ignored user @ !VlyOdAIoWKBBMwwrMU:moodle.com: @nextmoodlebot:moodle.com said 'Hey @stronk7:moodle.com!']
Fri, 30 Sep 2022 14:56:44 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Fri, 30 Sep 2022 14:56:44 GMT [INFO] [MatrixClientLite] Backing off for 10187.448193052402ms
Fri, 30 Sep 2022 14:56:55 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Fri, 30 Sep 2022 14:56:55 GMT [INFO] [MatrixClientLite] Backing off for 5569.095947478458ms
Fri, 30 Sep 2022 14:57:01 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Fri, 30 Sep 2022 14:57:01 GMT [INFO] [MatrixClientLite] Backing off for 9813.54110596994ms
Fri, 30 Sep 2022 14:57:11 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Fri, 30 Sep 2022 14:57:11 GMT [INFO] [MatrixClientLite] Backing off for 11588.864163629183ms
Fri, 30 Sep 2022 14:57:52 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Fri, 30 Sep 2022 14:57:52 GMT [INFO] [MatrixClientLite] Backing off for 11163.263685616239ms
Fri, 30 Sep 2022 14:58:04 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Fri, 30 Sep 2022 14:58:04 GMT [INFO] [MatrixClientLite] Backing off for 14792.47189950429ms
Fri, 30 Sep 2022 14:58:49 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
Fri, 30 Sep 2022 14:58:49 GMT [INFO] [MatrixClientLite] Backing off for 10313.26882274437ms
Fri, 30 Sep 2022 14:58:59 GMT [ERROR] [MatrixClientLite] Error handling sync Error: Expect value to be String, but received Undefined
...
```

Note that if I do exactly the same with the prod instance that is using the latest beta release, it's not exposed to the same problem.

Ciao :-)

**Comment by turt2live** [comment]
unfortunately for this issue we'd need to see the whole `/sync` request and response: Raising the log level to trace should do that, at risk of making the logs extremely noisy. `LogService` is the thing to call on that.

**Comment by stronk7** [comment]
Aha, oki, will give to it a try soon, thanks!

**Comment by stronk7** [comment]
Hi, here it's the log file with level set to `TRACE`. Basically what I did is:

1. Start the bot
2. Wait ~30 minutes
3. Send `!ping` message, the bot replied `Pong` and became stuck.

Hope it helps, ciao :-)

Attached: [trace.log](https://github.com/turt2live/matrix-bot-sdk/files/9689911/trace.log)

**Comment by turt2live** [comment]
circumstantially, it looks like the error is caused by sending a toDevice message from the bot (which happens internally). I suppose the OlmMachine is expecting a return value that we're not giving it or something.

Will look into it.

**Comment by stronk7** [comment]
For the records, and to discard factors, we have recently upgraded from synapse v1.57.1 to v1.68.0 and the same problem continues happening.

Ciao :-)

**Comment by turt2live** [comment]
I've opened an upstream issue on this as it does not appear like a bug in the bot-sdk: https://github.com/matrix-org/matrix-rust-sdk/issues/1093

I have however updated the rust-sdk bindings in hopes of magically fixing the issue (https://github.com/turt2live/matrix-bot-sdk/pull/266) - this should be available in the next release.

**Comment by stronk7** [comment]
Thanks for the update!

I've tried to bump here my dependencies to use:

```
 "matrix-bot-sdk": "turt2live/matrix-bot-sdk#main",
```

But for some reason (unknown to a npm/yarn newbie like me) that leads to all sort of errors here when `tsc` tries to compile it. So, surely, I'll wait till `0.6.3` comes out and will try then.

Ciao :-)

**Comment by stronk7** [comment]
Ah, I've just cleaned the lock file and, after `npm install` I can see in the new lock file that now I'm using

```
...
"node_modules/@matrix-org/matrix-sdk-crypto-nodejs": {
        "version": "0.1.0-beta.2",
...
```
(that is the new one, if I'm not wrong, so the beta1 => beta2 has happened, it seems.

Unfortunately, I've tried the usual routine:
1. Start the bot.
2. Wait some minutes.
3. Try some basic commands in unencrypted room, all ok.
4. Try `!hi` command in encrypted (1:1) room, I got the reply but also the `Expect value to be String, but received Undefined` and, from that point, 100% unresponsive.

Ciao :-)

**Comment by stronk7** [comment]
Ah, I've just cleaned the lock file and, after `npm install` I can see in the new lock file that now I'm using

```
...
"node_modules/@matrix-org/matrix-sdk-crypto-nodejs": {
        "version": "0.1.0-beta.2",
...
```
(that is the new one, if I'm not wrong, so the beta1 => beta2 has happened, it seems.

Unfortunately, I've tried the usual routine:
1. Start the bot.
2. Wait some minutes.
3. Try some basic commands in unencrypted room, all ok.
4. Try `!hi` command in encrypted (1:1) room, I got the reply but also the `Expect value to be String, but received Undefined` and, from that point, 100% unresponsive.

Ciao :-)

**Comment by stronk7** [comment]
I'm going to try with `matrix-bot-sdk v0.6.3` that comes with `matrix-sdk-crypto-nodejs 0.1.0-beta.3` in out staging infra... 🤞

**Comment by stronk7** [comment]
No luck! After upgrading from 0.6.2 to 0.6.3 I got a lot (one by encrypted room the bot is, i imagine) of:

```
Decryption error on !xxxxxxxxxx:xxxxx.com xxxxxxxxx[Error: decryption failed because the room key is missing] {
  code: 'GenericFailure'
}
```

So I went the brave key, discarded the session from element-web, stopped the bot, removed the crypto-storage completely and started again.

And exactly the very same behaviour reported here continues: immediately after getting the 1st response from the bot in an encrypted room, the `Expect value to be String, but received Undefined` begin, and it doesn't replie again ever.

Ciao :-)

**Comment by stronk7** [comment]
Hi,

sorry, forgot to comment earlier (Xmas and so on)... I've been running 0.6.3 since Xmas and have not got the error anymore. Last week I also updated the prod bot, that is in some dozens of rooms, plain and encrypted, and all right too.

So maybe this can be closed as fixed by 0.6.3 (not really sure which PR fixed it, heh).

Many thanks and ciao :-)

**Comment by turt2live** [comment]
awesome :)

We'll consider it fixed for now then - thanks!
